cocotb/types/logic.py

```diff
@@ -0,0 +1,207 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+from cocotb._py_compat import cache
+
+
+""" DESIGN NOTES
+
+__singleton_cache__
+    Used to ensure there is only ever one instance of a particular value.
+    This is a memory optimization: there will only ever be 4 objects allocated;
+    and also a speed optimization: value equality can be identity equality (as inherited from the `object` base class).
+
+operator `isinstance(other, type(self)`
+    Allows operation between a `Bit` and a `Logic` to return a `Logic`.
+
+strict type equality
+    To use `Logic` or `Bit` in hashable collections we need to define `__hash__` and `__eq__` such that
+    "Hashable objects which compare equal must have the same hash value."
+    Retrieved from https://hynek.me/articles/hashes-and-equality/ on 2021-03-20.
+    The best way to achieve this is to make `Logic` and `Bit` never equal, like `tuple` and `list`.
+    If we instead made them equal in hash and value, they would be substitutable; which we don't want.
+
+@cache
+    Shows extreme performance improvements.
+    Even faster than precomputing the results and storing them in the type.
+    Everything can be cached since there are a (hopefully) limited number of valid values in all subclasses.
+"""
+
+
+class Logic:
+    """
+    Model of a 4-value (``0``, ``1``, ``X``, ``Z``) datatype commonly seen in HDLs.
+
+    Modeled after (System)Verilog's 4-value ``logic`` type.
+    VHDL's 9-value ``std_ulogic`` type maps to this type by treating weak values as full strength values
+    and "uninitialized" (``U``) and "don't care" (``-``) as "unknown" (``X``).
+
+    Supports common logic operations ``&``, ``|``, ``^``, and ``~``.
+    Can be converted to and from :class:`int`, :class:`str`, :class:`bool`, and :class:`cocotb.types.Bit`.
+    """
+
+    __singleton_cache__ = {}
+
+    _repr_map = {
+        # 0 and weak 0
+        False: 0,
+        0: 0,
+        "0": 0,
+        "L": 0,
+        "l": 0,
+        # 1 and weak 1
+        True: 1,
+        1: 1,
+        "1": 1,
+        "H": 1,
+        "h": 1,
+        # unknown, unassigned, and weak unknown
+        None: 2,
+        "X": 2,
+        "x": 2,
+        "U": 2,
+        "u": 2,
+        "W": 2,
+        "w": 2,
+        "-": 2,
+        # high impedance
+        "Z": 3,
+        "z": 3,
+    }
+
+    @cache
+    def __new__(cls, value=None):
+        # convert to internal representation
+        try:
+            _repr = cls._repr_map[value]
+        except KeyError:
+            raise ValueError(
+                "{!r} is not convertable to a {}".format(value, cls.__qualname__)
+            ) from None
+        # ensure only one object is made per representation
+        if _repr not in cls.__singleton_cache__:
```

Comment:
Does the `@cache` above mean that this `if` is always true?

Reply:
No. The caches are different. This will fill the `__singleton_cache__` with the 4 values and ensure only those objects are returned, and the `@cache` will map the accepted convertible values to the value. Just using `@cache` is not sufficient. Instead of this line I could fill the cache after the class definition.

```python
class Logic:
    # ...

Logic.__singleton_cache__[0] = object.__new__(Logic)
Logic.__singleton_cache__[1] = object.__new__(Logic)
Logic.__singleton_cache__[2] = object.__new__(Logic)
Logic.__singleton_cache__[3] = object.__new__(Logic)
```

Reply:
Ok, that makes sense to me. I don't think filling the cache after really makes any difference, and it's clearer the way you have it.